Source/player.cpp

```diff
@@ -1683,152 +1683,109 @@ void __fastcall RespawnDeadItem(ItemStruct *itm, int x, int y)
+
+	if ( (DWORD)pnum >= 4 ) {
+		TermMsg("StartPlayerKill: illegal player %d", pnum);
+	}
+
+	if ( plr[pnum]._pClass == PC_WARRIOR ) {
+		PlaySfxLoc(PS_DEAD, plr[pnum].WorldX, plr[pnum].WorldY); // BUGFIX: should use `PS_WARR71` like other classes
+	} else if ( plr[pnum]._pClass == PC_ROGUE ) {
+		PlaySfxLoc(PS_ROGUE71, plr[pnum].WorldX, plr[pnum].WorldY);
+	} else if ( plr[pnum]._pClass == PC_SORCERER ) {
+		PlaySfxLoc(PS_MAGE71, plr[pnum].WorldX, plr[pnum].WorldY);
+	}
+
+	if ( plr[pnum]._pgfxnum ) {
+		plr[pnum]._pgfxnum = 0;
+		plr[pnum]._pGFXLoad = 0;
+		SetPlrAnims(pnum);
 	}
+
 	if ( !(plr[pnum]._pGFXLoad & PFILE_DEATH) ) {
 		LoadPlrGFX(pnum, PFILE_DEATH);
 	}
-	v6 = plr[v3 / 0x54D8]._pDWidth;
-	NewPlrAnim(v2, plr[0]._pDAnim[plr[v3 / 0x54D8]._pdir + v3 / 4], plr[v3 / 0x54D8]._pDFrames, 1, v6);
-	plr[v3 / 0x54D8]._pBlockFlag = 0;
-	plr[v3 / 0x54D8]._pmode = PM_DEATH;
-	plr[v3 / 0x54D8]._pInvincible = 1;
-	SetPlayerHitPoints(v2, 0);
-	v7 = v2 == myplr;
-	plr[v3 / 0x54D8]._pVar8 = 1;
-	if ( !v7 && !itm && !v20 )
-	{
-		v8 = &plr[v3 / 0x54D8].InvBody[0]._itype;
-		v9 = 7;
-		do
-		{
-			*v8 = -1;
-			v8 += 92;
-			--v9;
+
+	PlayerStruct *p = &plr[pnum];
+	NewPlrAnim(pnum, p->_pDAnim[plr[pnum]._pdir], p->_pDFrames, 1, p->_pDWidth);
+
+	plr[pnum]._pBlockFlag = FALSE;
+	plr[pnum]._pmode = PM_DEATH;
+	plr[pnum]._pInvincible = TRUE;
+	SetPlayerHitPoints(pnum, 0);
+	plr[pnum]._pVar8 = 1;
+
+	int i;
+	if ( pnum != myplr && !earflag && !diablolevel ) {
+		for ( i = 0; i < NUM_INVLOC; i++ ) {
+			plr[pnum].InvBody[i]._itype = ITYPE_NONE;
 		}
-		while ( v9 );
-		CalcPlrInv(v2, 0);
+		CalcPlrInv(pnum, 0);
 	}
-	if ( plr[v3 / 0x54D8].plrlevel == currlevel )
-	{
-		FixPlayerLocation(v2, plr[v3 / 0x54D8]._pdir);
-		RemovePlrFromMap(v2);
-		v10 = &dFlags[plr[v3 / 0x54D8].WorldX][plr[v3 / 0x54D8].WorldY];
-		*v10 |= 4u;
-		SetPlayerOld(v2);
-		if ( v2 == myplr )
-		{
+
+	if ( plr[pnum].plrlevel == currlevel ) {
+		FixPlayerLocation(pnum, plr[pnum]._pdir);
+		RemovePlrFromMap(pnum);
+		dFlags[plr[pnum].WorldX][plr[pnum].WorldY] |= 4;
+		SetPlayerOld(pnum);
+
+		if ( pnum == myplr ) {
 			drawhpflag = 1;
 			deathdelay = 30;
-			if ( pcurs >= CURSOR_FIRSTITEM )
-			{
-				PlrDeadItem(v2, &plr[v3 / 0x54D8].HoldItem, 0, 0);
+
+			if ( pcurs >= CURSOR_FIRSTITEM ) {
+				PlrDeadItem(pnum, &plr[pnum].HoldItem, 0, 0);
 				SetCursor(CURSOR_HAND);
 			}
-			if ( !v20 )
-			{
-				DropHalfPlayersGold(v2);
-				if ( itm != (struct ItemStruct *)-1 )
-				{
-					if ( itm )
-					{
+
+			if ( !diablolevel ) {
+				DropHalfPlayersGold(pnum);
+				if ( earflag != -1 ) {
+					if ( earflag != 0 ) {
+						ItemStruct ear;
 						SetPlrHandItem(&ear, IDI_EAR);
-						sprintf(ear._iName, "Ear of %s", plr[v3 / 0x54D8]._pName);
-						v11 = plr[v3 / 0x54D8]._pClass;
-						if ( v11 == 2 )
-						{
+						sprintf(ear._iName, "Ear of %s", plr[pnum]._pName);
+						if ( plr[pnum]._pClass == PC_SORCERER ) {
 							ear._iCurs = 19;

```

Comment:
Eventually, we should have an ICURS enum. For reference, the `item_graphic_id` enum of the notes repo is used for this, we could adapt it to devilution. Perhaps as `ICURS_FOO`.

https://github.com/sanctuary/notes/blob/master/include/enums.h#L396

Of specific note:

```c
	ITEM_GRAPHIC_ID_EAR_SORCEROR                = 19,
	ITEM_GRAPHIC_ID_EAR_WARRIOR                 = 20,
	ITEM_GRAPHIC_ID_EAR_ROGUE                   = 21,
```

```c
/// TODO: Add remaining enums for item_graphic_id.

/// Item graphic IDs; frame_num-11 of objcurs.cel.
typedef enum {
	ITEM_GRAPHIC_ID_POTION_OF_FULL_MANA         = 0,
	ITEM_GRAPHIC_ID_SCROLL_OF                   = 1,
	ITEM_GRAPHIC_ID_GOLD_SMALL                  = 4,
	ITEM_GRAPHIC_ID_GOLD_MEDIUM                 = 5,
	ITEM_GRAPHIC_ID_GOLD_LARGE                  = 6,
	ITEM_GRAPHIC_ID_RING_OF_TRUTH               = 10,
	ITEM_GRAPHIC_ID_RING                        = 12,
	ITEM_GRAPHIC_ID_SPECTRAL_ELIXIR             = 15,
	ITEM_GRAPHIC_ID_GOLDEN_ELIXIR               = 17,
	ITEM_GRAPHIC_ID_EMPYREAN_BAND               = 18,
	ITEM_GRAPHIC_ID_EAR_SORCEROR                = 19,
	ITEM_GRAPHIC_ID_EAR_WARRIOR                 = 20,
	ITEM_GRAPHIC_ID_EAR_ROGUE                   = 21,
	ITEM_GRAPHIC_ID_BLOOD_STONE                 = 25,
	ITEM_GRAPHIC_ID_ELIXIR_OF_VITALITY          = 31,
	ITEM_GRAPHIC_ID_POTION_OF_HEALING           = 32,
	ITEM_GRAPHIC_ID_POTION_OF_FULL_REJUVENATION = 33,
	ITEM_GRAPHIC_ID_ELIXIR_OF_MAGIC             = 34,
	ITEM_GRAPHIC_ID_POTION_OF_FULL_HEALING      = 35,
	ITEM_GRAPHIC_ID_ELIXIR_OF_DEXTERITY         = 36,
	ITEM_GRAPHIC_ID_POTION_OF_REJUVENATION      = 37,
	ITEM_GRAPHIC_ID_ELIXIR_OF_STRENGTH          = 38,
	ITEM_GRAPHIC_ID_POTION_OF_MANA              = 39,
	ITEM_GRAPHIC_ID_BRAIN                       = 40,
	ITEM_GRAPHIC_ID_OPTIC_AMULET                = 44,
	ITEM_GRAPHIC_ID_AMULET                      = 45,
	ITEM_GRAPHIC_ID_DAGGER                      = 51,
	ITEM_GRAPHIC_ID_BLADE                       = 56,
	ITEM_GRAPHIC_ID_BASTARD_SWORD               = 57,
	ITEM_GRAPHIC_ID_MACE                        = 59,
	ITEM_GRAPHIC_ID_LONG_SWORD                  = 60,
	ITEM_GRAPHIC_ID_BROAD_SWORD                 = 61,
	ITEM_GRAPHIC_ID_FALCHION                    = 62,
	ITEM_GRAPHIC_ID_MORNING_STAR                = 63,
	ITEM_GRAPHIC_ID_SHORT_SWORD                 = 64,
	ITEM_GRAPHIC_ID_CLAYMORE                    = 65,
	ITEM_GRAPHIC_ID_CLUB                        = 66,
	ITEM_GRAPHIC_ID_SABRE                       = 67,
	ITEM_GRAPHIC_ID_SPIKED_CLUB                 = 70,
	ITEM_GRAPHIC_ID_SCIMITAR                    = 72,
	ITEM_GRAPHIC_ID_FULL_HELM                   = 75,
	ITEM_GRAPHIC_ID_MAGIC_ROCK                  = 76,
	ITEM_GRAPHIC_ID_THE_UNDEAD_CROWN            = 78,
	ITEM_GRAPHIC_ID_HELM                        = 82,
	ITEM_GRAPHIC_ID_BUCKLER                     = 83,
	ITEM_GRAPHIC_ID_VIEL_OF_STEEL               = 85,
	ITEM_GRAPHIC_ID_BOOK_GREY                   = 86,
	ITEM_GRAPHIC_ID_BOOK_RED                    = 87,
	ITEM_GRAPHIC_ID_BOOK_BLUE                   = 88,
	ITEM_GRAPHIC_ID_BLACK_MUSHROOM              = 89,
	ITEM_GRAPHIC_ID_SKULL_CAP                   = 90,
	ITEM_GRAPHIC_ID_CAP                         = 91,
	ITEM_GRAPHIC_ID_HARLEQUIN_CREST             = 93,
	ITEM_GRAPHIC_ID_CROWN                       = 95,
	ITEM_GRAPHIC_ID_MAP_OF_THE_STARS            = 96,
	ITEM_GRAPHIC_ID_FUNGAL_TOME                 = 97,
	ITEM_GRAPHIC_ID_GREAT_HELM                  = 98,
	ITEM_GRAPHIC_ID_BATTLE_AXE                  = 101,
	ITEM_GRAPHIC_ID_HUNTERS_BOW                 = 102,
	ITEM_GRAPHIC_ID_FIELD_PLATE                 = 103,
	ITEM_GRAPHIC_ID_SMALL_SHIELD                = 105,
	ITEM_GRAPHIC_ID_CLEAVER                     = 106,
	ITEM_GRAPHIC_ID_STUDDED_LEATHER_ARMOR       = 107,
	ITEM_GRAPHIC_ID_SHORT_STAFF                 = 109,
	ITEM_GRAPHIC_ID_TWO_HANDED_SWORD            = 110,
	ITEM_GRAPHIC_ID_CHAIN_MAIL                  = 111,
	ITEM_GRAPHIC_ID_SMALL_AXE                   = 112,
	ITEM_GRAPHIC_ID_KITE_SHIELD                 = 113,
	ITEM_GRAPHIC_ID_SCALE_MAIL                  = 114,
	ITEM_GRAPHIC_ID_SHORT_BOW                   = 118,
	ITEM_GRAPHIC_ID_LONG_WAR_BOW                = 119,
	ITEM_GRAPHIC_ID_WAR_HAMMER                  = 121,
	ITEM_GRAPHIC_ID_MAUL                        = 122,
	ITEM_GRAPHIC_ID_LONG_STAFF                  = 123,
	ITEM_GRAPHIC_ID_WAR_STAFF                   = 124,
	ITEM_GRAPHIC_ID_TAVERN_SIGN                 = 126,
	ITEM_GRAPHIC_ID_HARD_LEATHER_ARMOR          = 127,
	ITEM_GRAPHIC_ID_RAGS                        = 128,
	ITEM_GRAPHIC_ID_QUILTED_ARMOR               = 129,
	ITEM_GRAPHIC_ID_FLAIL                       = 131,
	ITEM_GRAPHIC_ID_TOWER_SHIELD                = 132,
	ITEM_GRAPHIC_ID_COMPOSITE_BOW               = 133,
	ITEM_GRAPHIC_ID_GREAT_SWORD                 = 134,
	ITEM_GRAPHIC_ID_LEATHER_ARMOR               = 135,
	ITEM_GRAPHIC_ID_SPLINT_MAIL                 = 136,
	ITEM_GRAPHIC_ID_ROBE                        = 137,
	ITEM_GRAPHIC_ID_ANVIL_OF_FURY               = 140,
	ITEM_GRAPHIC_ID_BROAD_AXE                   = 141,
	ITEM_GRAPHIC_ID_LARGE_AXE                   = 142,
	ITEM_GRAPHIC_ID_GREAT_AXE                   = 143,
	ITEM_GRAPHIC_ID_AXE                         = 144,
	ITEM_GRAPHIC_ID_LARGE_SHIELD                = 147,
	ITEM_GRAPHIC_ID_GOTHIC_SHIELD               = 148,
	ITEM_GRAPHIC_ID_CLOAK                       = 149,
	ITEM_GRAPHIC_ID_CAPE                        = 150,
	ITEM_GRAPHIC_ID_FULL_PLATE_MAIL             = 151,
	ITEM_GRAPHIC_ID_GOTHIC_PLATE                = 152,
	ITEM_GRAPHIC_ID_BREAST_PLATE                = 153,
	ITEM_GRAPHIC_ID_RING_MAIL                   = 154,
	ITEM_GRAPHIC_ID_STAFF_OF_LAZARUS            = 155,
	ITEM_GRAPHIC_ID_ARKAINES_VALOR              = 157,
	ITEM_GRAPHIC_ID_SHORT_WAR_BOW               = 165,
	ITEM_GRAPHIC_ID_COMPOSITE_STAFF             = 166,
	ITEM_GRAPHIC_ID_SHORT_BATTLE_BOW            = 167,
	ITEM_GRAPHIC_ID_GOLD                        = 168,
} item_graphic_id;
```

Reply:
Enum added and used where applicable